ccf_transaction_processor/transaction_processor/pdo_tp.cpp

```diff
@@ -137,14 +163,139 @@ namespace ccfapp
+
+                    // Verify MREnclave
+                    std::vector<uint8_t> mrEnclaveFromReport_vector(mrEnclaveFromReport.m, mrEnclaveFromReport.m + SGX_HASH_SIZE);
+                    std::string mrEnclavFromReport_hex = ds::to_hex(mrEnclaveFromReport_vector);
+                    transform(mrEnclavFromReport_hex.begin(), mrEnclavFromReport_hex.end(), mrEnclavFromReport_hex.begin(), ::toupper);
+                    if (mrEnclavFromReport_hex != expected_measurements.mrenclave) {
+                        return ccf::make_error(HTTP_STATUS_BAD_REQUEST, ccf::errors::InvalidInput, "Enclave attestation report verification Failed. Invalid MREnclave");
+                    }
+
+                    // Verify Nonce
+                    if (in.registration_block_context != verification_report.nonce) {
+                        return ccf::make_error(
+                            HTTP_STATUS_BAD_REQUEST, ccf::errors::InvalidInput, "Enclave attestation report verification Failed. Invalid Nonce in the Verification Report");
+                    }
+
+                    // Verify Base Name
+                    std::vector<uint8_t> basenameFromReport_vector(mrBasename.name, mrBasename.name + BASENAME_SIZE);
+                    std::string basenameFromReport_hex = ds::to_hex(basenameFromReport_vector);
+                    transform(basenameFromReport_hex.begin(), basenameFromReport_hex.end(), basenameFromReport_hex.begin(), ::toupper);
+                    if (basenameFromReport_hex != expected_measurements.basename) {
+                        return ccf::make_error(
+                            HTTP_STATUS_BAD_REQUEST, ccf::errors::InvalidInput, "Enclave attestation report verification Failed. Invalid enclave base name");
+                    }
+
+                    // Verify user report data
+                    std::vector<uint8_t> originator_key(in.EHS_verifying_key.begin(), in.EHS_verifying_key.end());
+                    std::vector<uint8_t> originator_key_hash = crypto::SHA256(originator_key);
+                    std::string originator_key_hash_hex = ds::to_hex(originator_key_hash);
+                    originator_key_hash_hex.resize(ORIGINATOR_KEY_HASH_SIZE);
+                        //To understand why we truncate, check implementation of pdo.common.keys.ServiceKeys.hashed_indentity
+                    std::transform(originator_key_hash_hex.begin(), originator_key_hash_hex.end(), originator_key_hash_hex.begin(), ::tolower);
+
+                    std::string user_data_hash_input = in.verifying_key;
+                    user_data_hash_input += in.encryption_key;
+                    user_data_hash_input += originator_key_hash_hex;
+
+                    std::vector<uint8_t> user_data_hash_input_vector(user_data_hash_input.begin(), user_data_hash_input.end());
+                    std::vector<uint8_t> user_data_hash = crypto::SHA256(user_data_hash_input_vector);
+                    user_data_hash.resize(SGX_REPORT_DATA_SIZE, 0);
```

Comment:
I guess this is necessary because data hash is 32 bytes, but the sgx report is 64 (where the last 32 bytes are 0).
Maybe add a comment about that?

Reply:
thank you, good point, added.